std::string == ValAdaptor の比較ができないらしい

```
/Users/runner/work/webcface-tools/webcface-tools/tui/view.cc:279:27: error: invalid operands to binary expression ('std::string' and 'value_type' (aka 'webcface::ValAdaptor'))
            if (*bind_ref != selected_v) {
                ~~~~~~~~~ ^  ~~~~~~~~~~
```

--- a/src/include/webcface/common/val.h
+++ b/src/include/webcface/common/val.h
@@ -270,12 +270,19 @@ class ValAdaptor {
             return this->asStringRef() == other.asStringRef();
         }
     }
+    bool operator!=(const ValAdaptor &other) const { return !(*this == other); }
 
     template <typename T>
         requires(std::constructible_from<ValAdaptor, T> &&
                  !std::same_as<ValAdaptor, T>)
     bool operator==(const T &other) const {
         return *this == ValAdaptor(other);
+    }
+    template <typename T>
+        requires(std::constructible_from<ValAdaptor, T> &&
+                 !std::same_as<ValAdaptor, T>)
+    bool operator!=(const T &other) const {
+        return !(*this == other);
     }
 };
 

--- a/src/include/webcface/common/val.h
+++ b/src/include/webcface/common/val.h
@@ -270,21 +270,21 @@ class ValAdaptor {
             return this->asStringRef() == other.asStringRef();
         }
     }
-    bool operator!=(const ValAdaptor &other) const { return !(*this == other); }
 
     template <typename T>
         requires(std::constructible_from<ValAdaptor, T> &&
-                 !std::same_as<ValAdaptor, T>) bool
-    operator==(const T &other) const {
+                 !std::same_as<ValAdaptor, T>)
+    bool operator==(const T &other) const {
         return *this == ValAdaptor(other);
     }
-    template <typename T>
-        requires(std::constructible_from<ValAdaptor, T> &&
-                 !std::same_as<ValAdaptor, T>) bool
-    operator!=(const T &other) const {
-        return !(*this == other);
-    }
 };
+
+template <typename T>
+    requires(std::constructible_from<ValAdaptor, T> &&
+             !std::same_as<ValAdaptor, T>)
+bool operator==(const T &other, const ValAdaptor &val) {
+    return val == ValAdaptor(other);
+}
 
 inline std::ostream &operator<<(std::ostream &os, const ValAdaptor &a) {
     return os << static_cast<std::string>(a);

--- a/src/include/webcface/text.h
+++ b/src/include/webcface/text.h
@@ -169,8 +169,8 @@ class WEBCFACE_DLL Text : protected Field, public EventTarget<Text> {
      *
      */
     template <typename T>
-        requires std::same_as<T, Text> bool
-    operator==(const T &other) const {
+        requires std::same_as<T, Text>
+    bool operator==(const T &other) const {
         return static_cast<Field>(*this) == static_cast<Field>(other);
     }
     /*!
@@ -179,8 +179,8 @@ class WEBCFACE_DLL Text : protected Field, public EventTarget<Text> {
      *
      */
     template <typename T>
-        requires std::same_as<T, Text> bool
-    operator!=(const T &other) const {
+        requires std::same_as<T, Text>
+    bool operator!=(const T &other) const {
         return !(*this == other);
     }
     bool operator<(const Text &) const = delete;
@@ -305,17 +305,17 @@ class WEBCFACE_DLL InputRef {
     bool asBool() const { return get().asBool(); }
 
     template <typename T>
-        requires std::constructible_from<ValAdaptor, T> bool
-    operator==(const T &other) const {
+        requires std::constructible_from<ValAdaptor, T>
+    bool operator==(const T &other) const {
         return get() == other;
-    }
-    template <typename T>
-        requires std::constructible_from<ValAdaptor, T> bool
-    operator!=(const T &other) const {
-        return get() != other;
     }
 };
 
+template <typename T>
+    requires std::constructible_from<ValAdaptor, T>
+bool operator==(const T &other, const InputRef &ref) {
+    return ref == other;
+}
 inline std::ostream &operator<<(std::ostream &os, const InputRef &ref) {
     return os << ref.get();
 }

--- a/src/include/webcface/text.h
+++ b/src/include/webcface/text.h
@@ -314,7 +314,7 @@ class WEBCFACE_DLL InputRef {
 template <typename T>
     requires std::constructible_from<ValAdaptor, T>
 bool operator==(const T &other, const InputRef &ref) {
-    return ref == other;
+    return ref.get() == other;
 }
 inline std::ostream &operator<<(std::ostream &os, const InputRef &ref) {
     return os << ref.get();

--- a/src/include/webcface/text.h
+++ b/src/include/webcface/text.h
@@ -309,6 +309,11 @@ class WEBCFACE_DLL InputRef {
     bool operator==(const T &other) const {
         return get() == other;
     }
+    template <typename T>
+        requires std::constructible_from<ValAdaptor, T>
+    bool operator!=(const T &other) const {
+        return get() != other;
+    }
 };
 
 template <typename T>

--- a/src/test/func_test.cc
+++ b/src/test/func_test.cc
@@ -68,6 +68,13 @@ TEST_F(FuncTest, valAdaptor) {
     EXPECT_TRUE(ValAdaptor().empty());
     EXPECT_TRUE(ValAdaptor("").empty());
     EXPECT_TRUE(ValAdaptor(std::string("")).empty());
+
+    EXPECT_TRUE(ValAdaptor(1) == ValAdaptor(1.0));
+    EXPECT_TRUE(ValAdaptor(1) == 1);
+    EXPECT_TRUE(1 == ValAdaptor(1));
+    EXPECT_TRUE(ValAdaptor(1) == "1");
+    EXPECT_TRUE("1" == ValAdaptor(1));
+
 }
 TEST_F(FuncTest, field) {
     EXPECT_EQ(func("a", "b").member().name(), "a");

